File: app/src/main/java/com/example/myapplication/MainActivity.java
package com.example.myapplication;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import java.util.concurrent.ThreadLocalRandom;

@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
public class MainActivity extends AppCompatActivity {
    TextView score, orangeBox, blueBox, redBox, greenBox;
    int selectedNum = -1, randomNum = -1, result = 0;
    private static final long START_TIME_IN_MILLIS = Integer.MAX_VALUE;
    private CountDownTimer mCountDownTimer;
    private long mTimeLeftInMillis = START_TIME_IN_MILLIS;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findviews();
        startTimer();
        clickListener();

    }

    @SuppressLint("SetTextI18n")
    private void checkResult() {
        if (randomNum == selectedNum) {
            result++;
            score.setText(" " + result);
        }
        selectedNum = -1;
        resetTimer();

    }

    private void startTimer() {
        mCountDownTimer = new CountDownTimer(mTimeLeftInMillis, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                getOriginalState();
                getRandomNumber();
            }
            @Override
            public void onFinish() {
            }
        }.start();
    }

    private void resetTimer() {
        mCountDownTimer.cancel();
        startTimer();
    }

    private void getOriginalState() {
        orangeBox.setBackgroundColor(Color.YELLOW);
        blueBox.setBackgroundColor(Color.BLUE);
        redBox.setBackgroundColor(Color.RED);
        greenBox.setBackgroundColor(Color.GREEN);
    }

    private void setColor() {
        if (randomNum == 0) {
            orangeBox.setBackgroundColor(Color.parseColor("#dbdbdb"));

        } else if (randomNum == 1) {
            blueBox.setBackgroundColor(Color.parseColor("#dbdbdb"));

        } else if (randomNum == 2) {
            redBox.setBackgroundColor(Color.parseColor("#dbdbdb"));

        } else {
            greenBox.setBackgroundColor(Color.parseColor("#dbdbdb"));

        }
    }

    private void getRandomNumber() {
        randomNum = ThreadLocalRandom.current().nextInt(0, 4);
        setColor();
    }

    private void clickListener() {
        orangeBox.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                selectedNum = 0;
                checkResult();

            }
        });
        blueBox.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                selectedNum = 1;
                checkResult();

            }
        });
        redBox.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                selectedNum = 2;
                checkResult();

            }
        });
        greenBox.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                selectedNum = 3;
                checkResult();

            }
        });

    }

    private void findviews() {
        score = findViewById(R.id.score);
        orangeBox = findViewById(R.id.orangeBox);
        blueBox = findViewById(R.id.blueBox);
        redBox = findViewById(R.id.redBox);
        greenBox = findViewById(R.id.greenBox);
    }
}

File: app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.CountDownTimer
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.example.myapplication.databinding.ActivityMainBinding
import java.util.*

@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
class MainActivity : AppCompatActivity() {

    private var selectedNum = -1
    private var randomNum = -1
    private var resultVal = 0
    private var mCountDownTimer: CountDownTimer? = null
    private val mTimeLeftInMillis = START_TIME_IN_MILLIS
    private val TAG: String = "MainActivity"
    private lateinit var binding: ActivityMainBinding
    private var hasStarted = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        randomNumberAndResetColor
        clickListener()
    }

    @SuppressLint("SetTextI18n")
    private fun checkResult() {
        if(!hasStarted){
            startTimer()
            hasStarted = true
        }
        if (randomNum == selectedNum) {
            resultVal++
            binding.score.text = " $resultVal"

        } else {
            Log.d(TAG, "wrong touch")
            openResultPage()
        }
        selectedNum = -1
        resetTimer()


    }

    override fun onDestroy() {
        super.onDestroy()
        hasStarted = false
        mCountDownTimer?.cancel()
        Log.d(TAG, "on Destroy called")
    }

    private fun openResultPage() {
        val intent = Intent(this, ResultActivity::class.java)
        intent.putExtra("result", resultVal)
        finish()
        startActivity(intent)

    }

    private fun startTimer() {
        mCountDownTimer = object : CountDownTimer(mTimeLeftInMillis, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                originalState
                randomNumberAndResetColor
                Log.d(TAG, "tick$resultVal")
            }

            override fun onFinish() {
                Log.d(TAG, "timer finish")
                openResultPage()

            }
        }.start()
    }

    private fun resetTimer() {
        Log.d(TAG, "reset timer")
        mCountDownTimer!!.cancel()
        startTimer()
    }

    private val originalState: Unit
        get() {
            binding.orangeBox.setBackgroundColor(Color.YELLOW)
            binding.blueBox.setBackgroundColor(Color.BLUE)
            binding.redBox.setBackgroundColor(Color.RED)
            binding.greenBox.setBackgroundColor(Color.GREEN)
        }

    private fun setColor() {
        when (randomNum) {
            0 -> {
                binding.orangeBox.setBackgroundColor(Color.parseColor("#dbdbdb"))
            }
            1 -> {
                binding.blueBox.setBackgroundColor(Color.parseColor("#dbdbdb"))
            }
            2 -> {
                binding.redBox.setBackgroundColor(Color.parseColor("#dbdbdb"))
            }
            else -> {
                binding.greenBox.setBackgroundColor(Color.parseColor("#dbdbdb"))
            }
        }
    }

    private val randomNumberAndResetColor: Unit
        get() {
            getRandomNumber()
            setColor()
        }

    private fun getRandomNumber() {

        val randomNumTenp = Random().nextInt(4)
        if (randomNumTenp == randomNum) {
            getRandomNumber()
        } else {
            randomNum = randomNumTenp
            Log.d(TAG, "Random Num: $randomNum")

        }
    }

    private fun clickListener() {
        binding.orangeBox.setOnClickListener {
            selectedNum = 0
            checkResult()
        }
        binding.blueBox.setOnClickListener {
            selectedNum = 1
            checkResult()
        }
        binding.redBox.setOnClickListener {
            selectedNum = 2
            checkResult()
        }
        binding.greenBox.setOnClickListener {
            selectedNum = 3
            checkResult()
        }
    }

    companion object {
        private const val START_TIME_IN_MILLIS = 1000.toLong()
    }
}